include/net/http_client.h

```diff
@@ -0,0 +1,291 @@
+	 *
+	 * The reason-phrase element exists for the sole
+	 * purpose of providing a textual description
+	 * associated with the numeric status code. A client
+	 * SHOULD ignore the reason-phrase content.
+	 */
+	char http_status[HTTP_STATUS_STR_SIZE];
+
+	u8_t cl_present;
+	u8_t body_found;
+	u8_t message_complete;
+};
+
+/**
+ * Http context information. This contains all the data that is
+ * needed when working with http API.
+ */
+struct http_ctx {
+
+	/** socket id with http conenct*/
+	u16_t sock_id;
+
+	/** Server socket addr */
+	struct addrinfo *addr_info;
+	char host[HOST_NAME_MAX_SIZE];
+
+	/** HTTP request information */
+	struct http_request req;
+	char *req_buf;
+
+	/** HTTP response information */
+	struct http_resp rsp;
+	char *rsp_buf;
+
+	/** User provided HTTP response callback which is
+	 * called when a response is received to a sent HTTP
+	 * request.
+	 */
+	http_response_cb_t rsp_cb;
+
+	/** User provided to parse field-value what user care about
+	 * it, such as Content-Range to get the resource file size
+	 */
+	http_get_filed_value_cb_t fv_cb;
+
+	/** HTTP parser for parsing the initial request */
+	struct http_parser parser;
+
+	/** HTTP parser settings */
+	struct http_parser_settings parser_settings;
+};
+
+/**
+ * @brief Initialize user-supplied HTTP context.
+ *
+ * @param ctx HTTP context.
+ * @param host HTTP request host name.
+ * @param port http server port number.
+ * @param tls http connect is https or not
+ *
+ * @return Return 0 if ok, and <0 if error.
+ */
+int http_client_init(struct http_ctx *ctx,
+				 char *host,
+				 char *port,
+				 bool tls);
+
+/**
+ *@brief Generic function to get a HTTP request to the network.
+ *
+ * @param ctx HTTP context.
+ * @param path http request path, like /index.html.
+ * @param keep_alive http connect is alive or not
+ * @param user_data User data related to this request. This is passed
+ * to send callback if user has specified that.
+ *
+ * @return Return 0 if ok, and <0 if error.
+ */
+int http_client_get(struct http_ctx *ctx,
+		char *path,
+		bool keep_alive,
+		void *user_data);
+
+/**
+ * @brief Generic function to post a HTTP request to the network.
+ *
+ * @param ctx HTTP context.
+ * @param path http request path, like /index.html.
+ * @param keep_alive http connect is alive or not
+ * @param user_data User data related to this request. This is passed
+ * to send callback if user has specified that.
+ *
+ * @return Return 0 if ok, and <0 if error.
+ */
+int http_client_post(struct http_ctx *ctx,
+		char *path,
+		bool keep_alive,
+		void *user_data);
+
+/**
+ * @brief Close a network connection to peer.
+ *
+ * @param ctx Http context.
+ *
+ * @return 0 if ok, <0 if error.
+ */
+int http_client_close(struct http_ctx *ctx);
+
+/**
+ * @brief Add HTTP header field to the message.
+ *
+ * @details This can be called multiple times to add pieces of HTTP header
+ * fields into the message. Caller must put the "\r\n" characters to the
```

Comment:
The unterminated verbatim block error here is because ``\r`` is meaningful to doxygen comment processing.  Change this to be ``\\r\\n`` and the error should go away.

Reply:
well, thanks for your comment! it's done.